### DISCUSSION TITLE
How to implement a *.C file including a numerical integral function

### DISCUSSION THREAD

**Discussion by sSajjad90** [idaholab/moose#17258] [Q&A Modules: General] [answered]
Hi everyone,
I will define an integral to calculate the project's material properties, and I want to develop a customized *.C file. 
When I was exploring the available source files to find a sample, I found below user objects.
"ElementIntegralVariableUserObject", "ElementIntegralUserObject" but I do not know how these file  could be taken part in or even useful or not...
Would you please let me how I can implement the below integral? Somehow, with Or without using these user objects.

![2](https://user-images.githubusercontent.com/78385504/110500428-2f93e180-80c7-11eb-8129-93f7b061f5e3.JPEG)

I will solve the following equation at each time step to update "a1" for the other calculation. I got stock to how to figure out the integral. 
Any suggestion?

![1](https://user-images.githubusercontent.com/78385504/110500565-55b98180-80c7-11eb-8cfa-8443a278828c.JPEG)

**Comment by ngrilli** [comment, marked as answer]
Dear @sSajjad90 

The ElementIntegralVariableUserObject integrates a variable over the volume.
I think what you want to do is to integrate over the variable \rho; it is a one dimensional integral.

You could write your own custom material that calculates the integral using the coupled variable a_1
Then you store the integral in a material property
and you use the material property to implement a kernel for the second term in your last equation.

I think the most difficult part is calculating the integral, c++ probably does not have a function for that.
Otherwise, you will need to write your own .C code to integrate your function using an iterative method to check for convergence
because your integral goes to infinity.

Also, you can check if wolfram alpha gives you an analytical expression for your integral. 

Best Regards,
Nicolò Grilli
National University of Singapore
